### PR TITLE
[RyuJit/ARM32] Fix assertion of failure on "Instruction cannot be encoded"

### PR DESCRIPTION
**Pull request by sjsinju**
#10762, An assertion of failure has occurred during the conv_ovf test case is running.

I figured out the bit set mask for '0xffffff00' which is the imm is not supportted on ARM.
So at first, It would be checked the validation of using instruction with immediate constants.
If not, it would be use the instruction with registers.

### CODE REVIEW

**Comment by sjsinju** [comment]
```bash
sjsujinkim@sjsujinkim-rpi3:~/works/dotnet/Windows_NT.x64.Release/JIT/IL_Conformance/Old/Base/conv_ovf$ /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun conv_ovf.exe
Assert failure(PID 19978 [0x00004e0a], Thread: 19978 [0x4e0a]): Assertion failed 'tmpReg != REG_NA' in 'conv_ovf:u4_u1(int):int' (IL size 18)

    File: /home/sjsujinkim/works/dotnet/coreclr/src/jit/codegenarm.cpp Line: 2922
    Image: /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun

Aborted
```

**Comment by danmosemsft** [comment]
@dotnet-bot Test OSX10.12 x64 Checked Build and Test (build break fix)

**Comment by sjsinju** [comment]
@dotnet-bot Test Windows_NT x64 Debug Build and Test

**Comment by sjsinju** [comment]
I missed overflow check logics. And then I added it.
The result is
```
$ /home/sjsujinkim/works/dotnet/Windows_NT.x64.Release/Tests/coreoverlay/corerun conv_ovf.exe
$ echo $?
100
```

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib @dotnet/arm32-contrib

**Comment by sjsinju** [comment]
@dotnet-bot Test Please

**Comment by sjsinju** [comment]
@dotnet-bot help

**Comment by sjsinju** [comment]
@dotnet-bot test this please

**Comment by sjsinju** [comment]
@dotnet-bot Test OSX10.12 x64 Checked Build and Test
@dotnet-bot Test Ubuntu arm Cross Release Build 
@dotnet-bot Test Ubuntu x64 Checked Build and Test 
@dotnet-bot Test Ubuntu16.04 arm Cross Debug Build

File: src/jit/lsraarm.cpp [hunk 1]
```diff
@@ -330,10 +330,17 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
...
+                // If we cannot store data in an immediate for instructions,
+                // then we will need to reserve a temporary register.
 
+                // For checking the type
+                bool canStoreTypeMask = emitter::emitIns_valid_imm_for_alu(castInfo.typeMask);
+                if (!canStoreTypeMask)
+                {
+                    info->internalIntCount = 1;
+                }
+
```
Comment (BruceForstall) [inline src/jit/lsraarm.cpp]: It seems like the conditions here for specifying a temporary register don't quite replicate the conditions for using it in codegen. E.g., it seems this should be under
```
if (castInfo.unsignedSource || castInfo.unsignedDest)
```
or, better:
```
if (castInfo.signCheckOnly)
{
  // no temp needed
} else if (castInfo.unsignedSource || castInfo.unsignedDest)
  // check typeMask
} else {
  // check max/min value
}
```

**Comment by BruceForstall** [comment]
@sjsinju In addition to addressing my comments, you need to merge with my changes that merge arm32/arm64 code.

File: src/jit/lsraarm.cpp (hunk 1 above)
Comment (sjsinju) [inline src/jit/lsraarm.cpp]: Yes. It sounds better than before. Let me change.

**Comment by sjsinju** [comment]
@BruceForstall 
Why do below messages occur on all of x64 tests ?
```
Compile Error

1:17: error: use of undeclared identifier 'arm_Valid_Imm_For_Instr'
Indication 2 
```
Does 'codegenarmarch.cpp' compile on x64?

**Comment by BruceForstall** [comment]
@sjsinju Yes, `codegenarmarch.cpp` does compile on x64: we build a cross-compiling, x64-hosted, arm64-target "altjit" compiler. I haven't seen that error. I wonder why. This is on Linux x64? What is your environment? Is your environment different from the Jenkins CI testing that happens for a pull request?

**Comment by sjsinju** [comment]
@BruceForstall 'arm_Valid_Imm_For_Instr' is only defined on ARM like below
```cpp
#ifdef _TARGET_ARM_
    bool arm_Valid_Imm_For_Instr(instruction ins, ssize_t imm, insFlags flags);
    bool arm_Valid_Disp_For_LdSt(ssize_t disp, var_types type);
    bool arm_Valid_Imm_For_Alu(ssize_t imm);
    bool arm_Valid_Imm_For_Mov(ssize_t imm);
    bool arm_Valid_Imm_For_Small_Mov(regNumber reg, ssize_t imm, insFlags flags);
    bool arm_Valid_Imm_For_Add(ssize_t imm, insFlags flag);
    bool arm_Valid_Imm_For_Add_SP(ssize_t imm);
    bool arm_Valid_Imm_For_BL(ssize_t addr);

    bool ins_Writes_Dest(instruction ins);
#endif
```
I thought 'codegenarmarch.cpp' would not compile on other architectures. So I used 'arm_Valid_Imm_For_Instr' in this PR.
Actually, in my environment, it is compiled successfully because I compiled by cross for ARM. but as you can see CI testing, all of x64 tests were failed. This is no wonder.

**Comment by BruceForstall** [comment]
@sjsinju ok, I see the issue. Note that codegenarmarch.cpp is shared with both the arm32 and arm64 compilers. If you add code that is ARM32 specific, and does not apply to ARM64, you need to either put it under `#ifdef` or find some other way to make it work.

**Comment by sjsinju** [comment]
@BruceForstall Yes. So I added #ifdef as you said.

**Comment by BruceForstall** [comment]
@sjsinju You've changed behavior for arm64, which doesn't use a temp register (seems like it might also require the same changes you have made, but that can be done later). Instead, the code should probably look like:
```
#ifdef _TARGET_ARM_
            if (arm_Valid_Imm_For_Instr(INS_tst, castInfo.typeMask, INS_FLAGS_DONT_CARE))
            {
                emit->emitIns_R_I(INS_tst, cmpSize, sourceReg, castInfo.typeMask);
            }
            else
            {
                noway_assert(tmpReg != REG_NA);
                instGen_Set_Reg_To_Imm(cmpSize, tmpReg, castInfo.typeMask);
                emit->emitIns_R_R(INS_tst, cmpSize, sourceReg, tmpReg);
            }
#elif defined(_TARGET_ARM64_)
            emit->emitIns_R_I(INS_tst, cmpSize, sourceReg, castInfo.typeMask);
#endif
```

**Comment by BruceForstall** [comment]
@dotnet/arm64-contrib Does ARM64 require similar changes to those proposed here for ARM32?

**Comment by jashook** [comment]
@BruceForstall this test is passing on Arm64

File: src/jit/codegenarmarch.cpp [hunk 2]
```diff
@@ -1428,7 +1428,18 @@ void CodeGen::genIntToIntCast(GenTreePtr treeNode)
...
+                emit->emitIns_R_I(INS_tst, cmpSize, sourceReg, castInfo.typeMask);
+            }
+            else
+#endif // _TARGET_ARM
+            {
+                noway_assert(tmpReg != REG_NA);
+                instGen_Set_Reg_To_Imm(cmpSize, tmpReg, castInfo.typeMask);
+                emit->emitIns_R_R(INS_tst, cmpSize, sourceReg, tmpReg);
+            }
             emitJumpKind jmpNotEqual = genJumpKindForOper(GT_NE, CK_SIGNED);
```
Comment (sdmaclea) [inline src/jit/codegenarmarch.cpp]: This looks like it will break arm64, why is this not all in #ifdef.

**Comment by BruceForstall** [comment]
@jashook @sdmaclea @briansull Can arm64 TST immediate instruction encode all the `typeMask` immediates returned by `Lowering::getCastDescription()`? If so, it doesn't need the temp register version. But I don't have any intuition about arm64 bitmask immediates to know the answer.

**Comment by sdmaclea** [comment]
@BruceForstall There is no armv8 aarch64 tst instruction.  Why is this in codegenarmarch?

**Comment by sdmaclea** [comment]
@BruceForstall Please ignore my delerium I was looking at FP/SIMD instructions.

**Comment by sdmaclea** [comment]
@BruceForstall Looks like it can the aarch64 immediate is more flexible then the aaarch32/arm32 form.  It does look however that there are missing asserts in arm64 emitIns_R_I which should fire if the form could not be encoded.

**Comment by sdmaclea** [comment]
@BruceForstall I found the assert.  The function is just a bit inconsistent on where asserts are being placed.

**Comment by sjsinju** [comment]
@BruceForstall Yes. I was mistaken. I changed my code without looking at it properly.
It would be changed.

**Comment by sjsinju** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build

**Comment by BruceForstall** [comment]
I see:
```
19:35:38 FAILED   - JIT/Directed/coverage/importer/ldelemnullarr2/ldelemnullarr2.sh
19:35:38                BEGIN EXECUTION
19:35:38                /home/coreclr/bin/tests/Windows_NT.x64.Release/Tests/coreoverlay/corerun ldelemnullarr2.exe
19:35:38                qemu: uncaught target signal 4 (Illegal instruction) - core dumped
19:35:38                ./ldelemnullarr2.sh: line 200:    72 Illegal instruction     (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
19:35:38                Expected: 100
19:35:38                Actual: 132
19:35:38                END EXECUTION - FAILED
```

Is "Ubuntu arm Cross Release Build" expected to pass?

**Comment by sjsinju** [comment]
@BruceForstall I think all of PRs are met the same failure now.

**Comment by BruceForstall** [comment]
@dotnet/arm32-contrib @wateret @mmitche @parjong @RussKeldorph Who knows something about the "Ubuntu arm Cross Release Build" CI testing? It has been failing every PR for the last 18 hours. See https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm_cross_release_ubuntu_prtest/. All with a timeout after:
```
19:35:38 FAILED   - JIT/Directed/coverage/importer/ldelemnullarr2/ldelemnullarr2.sh
19:35:38                BEGIN EXECUTION
19:35:38                /home/coreclr/bin/tests/Windows_NT.x64.Release/Tests/coreoverlay/corerun ldelemnullarr2.exe
19:35:38                qemu: uncaught target signal 4 (Illegal instruction) - core dumped
19:35:38                ./ldelemnullarr2.sh: line 200:    72 Illegal instruction     (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
19:35:38                Expected: 100
19:35:38                Actual: 132
19:35:38                END EXECUTION - FAILED
```

It was curious to me that the first failure was due to https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm_cross_release_ubuntu_prtest/1089/, for a PR that changes the Linux test build drastically. I presume the CI build/test machines get cleaned, and don't accidentally pick up changes left behind by failing PR testing?

File: src/jit/codegenarmarch.cpp (hunk 2 above)
Comment (sdmaclea) [inline src/jit/codegenarmarch.cpp]: Thanks.  Also, my apologizing for missing the fact that @BruceForstall had already commented on this.

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build

**Pull request merged**